agent/grpc-external/services/resource/write.go

```diff
@@ -0,0 +1,168 @@
+package resource
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/oklog/ulid/v2"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/hashicorp/consul/internal/storage"
+	"github.com/hashicorp/consul/lib/retry"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+func (s *Server) Write(ctx context.Context, req *pbresource.WriteRequest) (*pbresource.WriteResponse, error) {
+	if err := validateWriteRequiredFields(req); err != nil {
+		return nil, err
+	}
+
+	reg, err := s.resolveType(req.Resource.Id.Type)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check the user sent the correct type of data.
+	if !req.Resource.Data.MessageIs(reg.Proto) {
+		got := strings.TrimPrefix(req.Resource.Data.TypeUrl, "type.googleapis.com/")
+
+		return nil, status.Errorf(
+			codes.InvalidArgument,
+			"resource.data is of wrong type (expected=%q, got=%q)",
+			reg.Proto.ProtoReflect().Descriptor().FullName(),
+			got,
+		)
+	}
+
+	// At the storage backend layer, all writes are CAS operations.
+	//
+	// This makes it possible to *safely* do things like keeping the Uid stable
+	// across writes, carrying statuses over, and passing the current version of
+	// the resource to hooks, without restricting ourselves to only using the more
+	// feature-rich storage systems that support "patch" updates etc. natively.
+	//
+	// Although CAS semantics are useful for machine users like controllers, human
+	// users generally don't need them. If the user is performing a non-CAS write,
+	// we read the current version, and automatically retry if the CAS write fails.
+	var result *pbresource.Resource
+	err = s.retryCAS(ctx, req.Resource.Version, func() error {
+		input := clone(req.Resource)
+
+		// We read with EventualConsistency here because:
+		//
+		//	- In the common case, individual resources are written infrequently, and
+		//	  when using the Raft backend followers are generally within a few hundred
+		//	  milliseconds of the leader, so the first read will probably return the
+		//	  current version.
+		//
+		//	- StrongConsistency is expensive. In the Raft backend, it involves a round
+		//	  of heartbeats to verify cluster leadership (on top of log replication).
+		//
+		//	- CAS failures will be retried anyway. So the read-modify-write cycle
```

Comment:
minor doc: Just so the reader doesn't get confused with the "retried" mentioned here and the s.retryCAS(..) function, a little tweak might be useful:

- CAS failures will be retried **by the caller** anyway.

Reply:
Thanks for the call out! Retried here does actually refer to `retryCAS` 😅 so this could definitely use a little clarifying.